Source/ASNetworkImageNode.mm

```diff
@@ -773,18 +785,31 @@ - (void)_lazilyLoadImageIfNecessary
           }
           
           if ([imageContainer asdk_image] == nil && self->_downloader != nil) {
+            if (self->_delegateFlags.delegateWillLoadImageFromNetwork) {
+              [delegate imageNodeWillLoadImageFromNetwork:self];
```

Comment:
All of this methods are called with the lock held. This can cause deadlocks. We should at least document this behavior if we cannot get around this.

Reply:
I don't think these are locked. I tried putting a ASUnlockScope which return in an assertion failure.

Reply:
This is inside a completion block and the lock is no longer held by the time it's executed. That's why we grabbed the lock before checking the cache sentinel a few line above. So we don't need to unlock here, *but* we need to grab the lock before accessing `_delegateFlags`. We can do so at the beginning of this method, similar to how we the `_delegateFlags.delegateDidStartFetchingData` flag.